### PR TITLE
Fix/issue 94

### PR DESCRIPTION
**Pull request by Vassyli**
Fixes issue 94 by introducing a new series of classes, EventContext and the immutable EventContextData.

EventContext contains EventContextData, as well as information about the event (event name and matching pattern for this event). `handleEvent` is now expected to accept only `EventContext` and must return it; `publish` expects now an event name and a `EventContextDataContainer` instance and returns the `EventContextDataContainer`.

Examples:
```php
$contextData = NewViewpoint::create([
    'character' => $this->getCharacter(),
    'scene' => null
]);

$contextData = $this->getEventManager()->publish('h/lotgd/core/default-scene', $contextData);

$s = $contextData->get("scene");
```

```php
public static function handleEvent(Game $g, EventContext $context): EventContext
{
    if ($context->getEvent() === "h/lotgd/core/default-scene") {
        if (!$context->hasDataType(NewViewpoint::class)) {
            throw new \Exception(sprintf(
                "Context was expected to be %s, %s instead.",
                NewViewpoint::class,
                get_class($context->getData())
            ));
        }

        $context->setDataField("scene", $g->getEntityManager()->getRepository(Scene::class)->find(1));
    }
    return $context;
}
```

### CODE REVIEW

File: src/EventManager.php [hunk 1]
```diff
@@ -49,9 +52,17 @@ public function publish(string $event, array &$context)
             if (preg_match($s->getPattern(), $event)) {
                 $class = $s->getClass();
                 $this->g->getLogger()->addDebug("  Handling with {$class}.");
-                $c = $class::handleEvent($this->g, $event, $context);
+
+                $eventContext = new EventContext($event, $s->getPattern(), $contextData);
+
+                $returnedEventContext = $class::handleEvent($this->g, $eventContext);
+                if ($returnedEventContext->hasDataChanged($contextData)) {
```
Comment (austenmc) [inline src/EventManager.php]: Why does this if matter? Why not just write over `$contextData` with `$returnedEventContext->getData()` no matter what?

File: src/Events/NavigateToScene.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Fields are:
+ *  referrer    Scene|null
+ *  viewpoint   Viewpoint
+ *  scene       Scene
+ *  parameters  array
+ *  redirect    Scene|null
+ * @package LotGD\Core\Events
+ */
+class NavigateToScene extends EventContextDataContainer
```
Comment (austenmc) [inline src/Events/NavigateToScene.php]: Let's call these objects `NavigateToSceneData` etc. I know it's long but I was confused by the verb names in your example. Having the noun attached to the end makes it consistent with the superclass and clearer when reading.

File: src/Events/NewViewpoint.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * NewViewpoint data container which is used if no scene has ever been visited.
+ *
+ * Fields are:
+ *  character   Character
+ *  scene       Scene|null
+ * @package LotGD\Core\Events
+ */
+class NewViewpoint extends EventContextDataContainer
```
Comment (austenmc) [inline src/Events/NewViewpoint.php]: `NewViewpointData`

File: src/Game.php [hunk 4]
```diff
@@ -291,17 +294,18 @@ private function navigateToScene(Scene $scene, array $parameters)
             // Let and installed listeners (ie modules) make modifications to the
             // new viewpoint, including the ability to redirect the user to
             // a different scene, by setting $context['redirect'] to a new scene.
-            $context = [
+            $contextData = NavigateToScene::create([
```
Comment (austenmc) [inline src/Game.php]: Code like this is confusing when `NavigateToScene` is used as a noun.

File: src/Events/EventContextDataContainer.php [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+use LotGD\Core\Exceptions\ArgumentException;
+
+/**
+ * Abstract EventContextDataContainer to provide a basic structure for managing contextual data of an event.
+ *
+ * This class must be immutable and returns always a new instance of itself for any change.
+ * @package LotGD\Core\Events
+ * @immutable
+ */
+abstract class EventContextDataContainer
```
Comment (austenmc) [inline src/Events/EventContextDataContainer.php]: Why make this abstract and then make a completely empty subclass called `EventContextData`? I would just make this a regular class called `EventContextData`.

**Review by austenmc — APPROVED** [review]
Approved but let me know if you disagree w/ the naming suggestions. I think they make things cleaner.

Thanks for doing this!

File: src/EventManager.php (hunk 1 above)
Comment (Vassyli) [reply src/EventManager.php]: You're right, it doesn't matter at all.

**Pull request merged**